agent/hcp/telemetry/gauge_store_test.go

```diff
@@ -0,0 +1,81 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+func TestGaugeStore(t *testing.T) {
+	t.Parallel()
+
+	gaugeStore := NewGaugeStore()
+
+	attributes := []attribute.KeyValue{
+		{
+			Key:   attribute.Key("test_key"),
+			Value: attribute.StringValue("test_value"),
+		},
+	}
+
+	gaugeStore.Set("test", 1.23, attributes)
+
+	// Should store a new gauge.
+	val, ok := gaugeStore.LoadAndDelete("test")
+	require.True(t, ok)
+	require.Equal(t, val.Value, 1.23)
+	require.Equal(t, val.Attributes, attributes)
+
+	// Gauge with key "test" have been deleted.
+	val, ok = gaugeStore.LoadAndDelete("test")
+	require.False(t, ok)
+	require.Nil(t, val)
+
+	gaugeStore.Set("duplicate", 1.5, nil)
+	gaugeStore.Set("duplicate", 6.7, nil)
+
+	// Gauge with key "duplicate" should hold the latest (last seen) value.
+	val, ok = gaugeStore.LoadAndDelete("duplicate")
+	require.True(t, ok)
+	require.Equal(t, val.Value, 6.7)
+}
+
+func TestGaugeCallback_Failure(t *testing.T) {
+	t.Parallel()
+
+	k := "consul.raft.apply"
+	gaugeStore := NewGaugeStore()
+	gaugeStore.Set(k, 1.23, nil)
+
+	cb := gaugeStore.gaugeCallback(k)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	cancel()
+	err := cb(ctx, nil)
+	require.ErrorIs(t, err, context.Canceled)
+}
+
+// TestGaugeStore_Race induces a race condition. When run with go test -race,
+// this test should pass if  implementation is concurrency safe.
+func TestGaugeStore_Race(t *testing.T) {
+	t.Parallel()
+
+	gaugeStore := NewGaugeStore()
+	wg := &sync.WaitGroup{}
+	for k, v := range map[string]float64{"consul.raft.apply": 23.23, "consul.raft.test": 14.3} {
```

Comment:
Could we increase the test size from two samples ? Would a function that would generate 100 k,v would give us a better representation ? Otherwise two competing gorountines isn't likely to reveal much about the concurrency issues this might face.

Reply:
ah yeah I was wondering about that and fuzz testing but decided to keep it simple. I'll update to 100 generated kvs

Reply:
https://go.dev/blog/race-detector

> Because of its design, the race detector can detect race conditions only when they are actually triggered by running code, which means it’s important to run race-enabled binaries under realistic workloads. 

I might be wrong but I thought you still needed realistic workloads.

Reply:
hmm that makes sense actually, yeah but still more realistic than 2 🤔 I'll update the PR, thanks for the context. I was misunderstanding the way it works I think!

Reply:
>which should theoretically be able to tell from just 2 operations

As @chapmanc said, unfortunately 2 ops is significantly less likely to perform a race due to the optimizations the OS makes at runtime. Using 100+ random generated results would be more reliable.

Reply:
thank you both, TIL. I'm curious and will look more into the race detector for fun too. :D I updated the PR to have 100 samples for the gauge kv